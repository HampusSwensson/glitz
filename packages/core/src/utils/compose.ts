export function compose<T>(...fns: Array<(arg: T) => T>) {
  return fns.reduceRight((prevFn, nextFn) => arg => nextFn(prevFn(arg)));
}
