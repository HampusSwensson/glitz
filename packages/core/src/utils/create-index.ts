type Index<TValue = string> = Record<string, TValue>;

export type InjectorIndexes = [
  getIndex: (selector?: string | undefined, condition?: string | undefined) => Index<string>,
  keyframesIndex: Index,
  fontFaceIndex: string[],
  clone: () => InjectorIndexes,
  plainIndex: Index,
  plainSupportsIndex: Index<Index>,
  selectorIndex: Index<Index>,
  selectorSupportsIndex: Index<Index<Index>>,
];

export function createInjectorIndexes(
  plainIndex: Index = {},
  plainSupportsIndex: Index<Index> = {},
  selectorIndex: Index<Index> = {},
  selectorSupportsIndex: Index<Index<Index>> = {},
  keyframesIndex: Index = {},
  fontFaceIndex: string[] = [],
): InjectorIndexes {
  return [
    (selector?: string, condition?: string) => {
      let index = plainIndex;

      if (condition) {
        if (selector) {
          selectorSupportsIndex[condition] ??= {};
          selectorSupportsIndex[condition][selector] ??= {};
          index = selectorSupportsIndex[condition][selector];
        } else {
          plainSupportsIndex[condition] ??= {};
          index = plainSupportsIndex[condition];
        }
      } else if (selector) {
        selectorIndex[selector] ??= {};
        index = selectorIndex[selector];
      }

      return index;
    },
    keyframesIndex,
    fontFaceIndex,
    () =>
      createInjectorIndexes(
        clone(plainIndex),
        clone(plainSupportsIndex),
        clone(selectorIndex),
        clone(selectorSupportsIndex),
        clone(keyframesIndex),
        fontFaceIndex.slice(0),
      ),
    plainIndex,
    plainSupportsIndex,
    selectorIndex,
    selectorSupportsIndex,
  ];
}

function clone<TIndex extends Index<string | Index | Index<Index>>>(index: TIndex): TIndex {
  const copy: Index<string | Index | Index<Index>> = {};

  for (const key in index) {
    const value: string | Index | Index<Index> = index[key];
    if (typeof value === 'string') {
      copy[key] = value;
    } else {
      copy[key] = clone(value);
    }
  }

  return copy as TIndex;
}
