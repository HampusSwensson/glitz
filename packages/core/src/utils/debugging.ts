type AnyObject = { [property: string]: any };
export let issueFormatter: (
  message: string,
  object: AnyObject,
  highlight?: AnyObject,
  replacer?: ((key: string, value: any) => any) | null,
  transform?: (issueObject: AnyObject) => AnyObject,
) => string[] = () => {
  return [];
};

if (process.env.NODE_ENV !== 'production') {
  const issueValue = '%o';

  issueFormatter = (message, object, highlight = {}, replacer = null, transform) => {
    const issueArgs = [];
    let issueObject: any = {};

    for (const key in object) {
      const isIssue = key in highlight;

      issueObject[isIssue ? `%c${key}%c` : key] = isIssue ? issueValue : object[key];

      if (isIssue) {
        issueArgs.push(
          'font-weight: bold; text-decoration: underline',
          'font-weight: normal; text-decoration: none',
          object[key],
        );
      }
    }

    if (transform) {
      issueObject = transform(issueObject);
    }

    let block = JSON.stringify(issueObject, replacer as any, 2);

    for (const key in highlight) {
      block = block.replace(`  "%c${key}%c": "${issueValue}"`, `  %c"${key}"%c: ${issueValue}`);
    }

    return [`${message}\n\n${block}`, ...issueArgs];
  };
}
