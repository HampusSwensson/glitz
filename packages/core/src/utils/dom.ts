export function injectSheetRule(styleElement: HTMLStyleElement, rule: string) {
  const sheet = styleElement.sheet as CSSStyleSheet;
  const index = sheet.cssRules.length;
  sheet.insertRule(rule, index);
  return sheet.cssRules[index] as CSSStyleRule;
}

export function createStyleElement(media?: string | null) {
  const element = document.createElement('style');

  if (media) {
    element.media = media;
  }

  return element;
}

export function insertStyleElement(element: HTMLStyleElement, insertBefore: HTMLStyleElement | null) {
  document.head.insertBefore(element, insertBefore);
  return element;
}
