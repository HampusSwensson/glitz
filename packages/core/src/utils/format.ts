export function formatRule(identifier: string, block: string) {
  return `${identifier}{${block}}`;
}

export function formatClassRule(className: string, block: string, selector: string = '') {
  return `.${formatRule(className + selector, block)}`;
}

export function formatKeyframesRule(name: string, blockList: string) {
  return `@keyframes ${formatRule(name, blockList)}`;
}

export function formatFontFaceRule(block: string) {
  return `@font-face {${block}}`;
}

const PRETTY_REGEX = /[{:;}]|(?:(["']).*?\1)/g;

export function prettifyRule(rule: string) {
  return rule.replace(PRETTY_REGEX, match =>
    match === '{' ? ' {\n  ' : match === ':' ? ': ' : match === ';' ? ';\n  ' : match === '}' ? ';\n}' : match,
  );
}
