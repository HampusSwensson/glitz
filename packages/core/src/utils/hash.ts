export type HashCounter = {
  (): string;
  clone(): HashCounter;
};

export function createHashCounter(prefix = '', count = 0, offset = 10, msb = 35, power = 1): HashCounter {
  function increment(): string {
    const virtualCount = count + offset;

    if (virtualCount === msb) {
      offset += (msb + 1) * 9;
      msb = Math.pow(36, ++power) - 1;
    }

    count++;

    // Skip "ad" due to ad-blockers
    if (virtualCount === 373) {
      return increment();
    }

    return prefix + virtualCount.toString(36);
  }

  return Object.assign(increment, {
    clone() {
      return createHashCounter(prefix, count, offset, msb, power);
    },
  });
}
