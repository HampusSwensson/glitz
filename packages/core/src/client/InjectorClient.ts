// tslint:disable no-conditional-assignment

import Injector from '../core/Injector';
import { injectSheetRule, isCSSStyleSheet } from '../utils/dom';
import { formatClassRule, formatFontFaceRule, formatKeyframesRule } from '../utils/format';

const CLASS_RULE_REGEX = /\.([a-z0-9]+)(:[^{]+)?\{([^}]+)\}/g;
const KEYFRAMES_REGEX = /@keyframes ([a-z0-9]+)\{((?:[a-z0-9%]+\{[^}]+\})+)\}/g;
const FONT_FACE_REGEX = /@font-face \{(.+?);font-family:([^}]+)\}/g;

export default class InjectorClient extends Injector {
  constructor(
    styleElement: HTMLStyleElement,
    incrementClassHash: () => string,
    incrementKeyframesHash: () => string,
    incrementFontFaceHash: () => string,
  ) {
    const plainIndex: { [block: string]: string } = {};
    const pseudoIndex: { [pseudo: string]: { [block: string]: string } } = {};
    const keyframesIndex: { [blockList: string]: string } = {};
    const fontFaceIndex: { [block: string]: string } = {};

    // Hydrate
    const css = styleElement.textContent;
    if (css) {
      let rule: RegExpExecArray | null;
      while ((rule = CLASS_RULE_REGEX.exec(css))) {
        incrementClassHash();
        const index = rule[2] ? (pseudoIndex[rule[2]] = pseudoIndex[rule[2]] || {}) : plainIndex;
        index[rule[3]] = rule[1];
      }
      while ((rule = KEYFRAMES_REGEX.exec(css))) {
        incrementKeyframesHash();
        keyframesIndex[rule[2]] = rule[1];
      }
      while ((rule = FONT_FACE_REGEX.exec(css))) {
        incrementFontFaceHash();
        fontFaceIndex[rule[1]] = rule[2];
      }
    }

    const sheet = styleElement.sheet;

    if (!isCSSStyleSheet(sheet)) {
      throw new Error('HTMLStyleElement was not inserted properly into DOM');
    }

    const injectNewClassRule = (className: string, block: string, pseudo?: string) => {
      const rule = formatClassRule(className, block, pseudo);
      injectSheetRule(sheet, rule);
    };

    const injectNewKeyframesRule = (name: string, blockList: string) => {
      const rule = formatKeyframesRule(name, blockList);
      injectSheetRule(sheet, rule);
    };

    const injectNewFontFaceRule = (name: string, block: string) => {
      const rule = formatFontFaceRule(name, block);
      injectSheetRule(sheet, rule);
    };

    super(
      plainIndex,
      pseudoIndex,
      keyframesIndex,
      fontFaceIndex,
      incrementClassHash,
      incrementKeyframesHash,
      incrementFontFaceHash,
      injectNewClassRule,
      injectNewKeyframesRule,
      injectNewFontFaceRule,
    );
  }
}
