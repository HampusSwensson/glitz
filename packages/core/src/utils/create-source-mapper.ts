import { createHashCounter } from './hash';

export type SourceMapper = (stackIndex: number, stackInfo: {}) => string;

export function createSourceMapper(prefix: string): SourceMapper {
  const blob = new Blob([`importScripts("https://unpkg.com/css-to-js-sourcemap-worker@2.0.4/worker.js")`], {
    type: 'application/javascript',
  });

  const worker = new Worker(URL.createObjectURL(blob));

  worker.postMessage({
    id: 'init_wasm',
    url: 'https://unpkg.com/css-to-js-sourcemap-worker@2.0.4/mappings.wasm',
  });

  worker.postMessage({
    id: 'set_render_interval',
    interval: 120,
  });

  // if (module && (module as any).hot && (module as any).hot.addStatusHandler) {
  //   (module as any).hot.addStatusHandler((status: string) => {
  //     if (status === 'dispose') {
  //       worker.postMessage({ id: 'invalidate' });
  //     }
  //   });
  // }

  worker.onmessage = msg => {
    const { id, css } = msg.data;
    if (id === 'render_css' && css) {
      const style = document.createElement('style');
      style.appendChild(document.createTextNode(css));
      (document.head as HTMLHeadElement).appendChild(style);
    }
  };

  const debugClassName = createDebugClassName(prefix);

  return (stackIndex, stackInfo) => {
    const className = debugClassName();

    worker.postMessage({
      id: 'add_mapped_class',
      className,
      stackInfo,
      stackIndex,
    });

    return className;
  };
}

export function createDebugClassName(prefix: string) {
  const debugHasher = createHashCounter(prefix);

  return () => {
    return `${prefix}debug-${debugHasher()}`;
  };
}
