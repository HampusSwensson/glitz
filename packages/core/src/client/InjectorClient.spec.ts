import { createHashCounter } from '../utils/hash';
import InjectorClient from './InjectorClient';

beforeEach(() => {
  document.head.innerHTML = '';
});

describe('client', () => {
  it('injects plain rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(injector.injectClassName({ color: 'red' })).toBe('a');
    expect(injector.injectClassName({ color: 'green', backgroundColor: 'black' })).toBe('b');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(2);
    expect(sheet.cssRules[0].cssText).toMatchSnapshot();
    expect(sheet.cssRules[1].cssText).toMatchSnapshot();
  });
  it('injects pseudo rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(injector.injectClassName({ color: 'red' }, ':hover')).toBe('a');
    expect(injector.injectClassName({ color: 'green', backgroundColor: 'black' }, ':hover')).toBe('b');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(2);
    expect(sheet.cssRules[0].cssText).toMatchSnapshot();
    expect(sheet.cssRules[1].cssText).toMatchSnapshot();
  });
  it('injects keyframes rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(injector.injectKeyframes({ from: { color: 'red' }, to: { color: 'green' } })).toBe('a');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(1);
    expect(sheet.cssRules[0].cssText).toMatchSnapshot();
  });
  it('injects font face rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(
      injector.injectFontFace({
        fontStyle: 'normal',
        fontWeight: 400,
        src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
      }),
    ).toBe('a');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(1);
    expect(sheet.cssRules[0].cssText).toMatchSnapshot();
  });
  it('injects fallback rule', () => {
    // Use Puppeteer
  });
  it('reuses plain rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(injector.injectClassName({ color: 'red' })).toBe('a');
    expect(injector.injectClassName({ color: 'red' })).toBe('a');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(1);
  });
  it('reuses pseudo rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(injector.injectClassName({ color: 'red' }, ':hover')).toBe('a');
    expect(injector.injectClassName({ color: 'red' }, ':hover')).toBe('a');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(1);
  });
  it('reuses keyframes rule', () => {
    const style = createStyle();
    const injector = createInjector(style);

    expect(injector.injectKeyframes({ from: { color: 'red' }, to: { color: 'green' } })).toBe('a');
    expect(injector.injectKeyframes({ from: { color: 'red' }, to: { color: 'green' } })).toBe('a');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules).toHaveLength(1);
  });
  it('reuses fallback rule', () => {
    // Use Puppeteer
  });
  it('hydrates plain rule', () => {
    const style = createStyle('.a{color:red}.b{color:green}.c{color:black;background-color:white}');
    const injector = createInjector(style);

    // Skipping .a
    expect(injector.injectClassName({ color: 'green' })).toBe('b');
    expect(injector.injectClassName({ color: 'black', backgroundColor: 'white' })).toBe('c');
  });
  it('hydrates pseudo rule', () => {
    const style = createStyle('.a:hover{color:red}.b:hover{color:green}');
    const injector = createInjector(style);

    // Skipping .a
    expect(injector.injectClassName({ color: 'green' }, ':hover')).toBe('b');
  });
  it('hydrates keyframes rule', () => {
    const style = createStyle(
      '@keyframes a{from{color:red}to{color:green}}@keyframes b{from{color:black}to{color:white}}',
    );
    const injector = createInjector(style);

    // Skipping .a
    expect(injector.injectKeyframes({ from: { color: 'black' }, to: { color: 'white' } })).toBe('b');
  });
  it('hydrates font face rule', () => {
    const style = createStyle(
      "@font-face {font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:a}" +
        "@font-face {font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:b}",
    );
    const injector = createInjector(style);

    // Skipping .a
    expect(
      injector.injectFontFace({
        fontStyle: 'normal',
        fontWeight: 400,
        src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
      }),
    ).toBe('b');
  });
  it('hydrates fallback rule', () => {
    const style = createStyle('.a{color:red;color:green}.b{color:black;color:white}');
    const injector = createInjector(style);

    // Skipping .a
    expect(injector.injectClassName({ color: ['black', 'white'] })).toBe('b');
  });
  it('increments plain hash', () => {
    const style = createStyle('.a{color:red}.b{color:green}');
    const injector = createInjector(style);

    expect(injector.injectClassName({ color: 'blue' })).toBe('c');

    const sheet = style.sheet as CSSStyleSheet;

    expect(sheet.cssRules[2].cssText).toMatchSnapshot();
  });
});

function createStyle(css?: string) {
  const element = document.createElement('style');
  document.head.appendChild(element);

  if (css) {
    element.appendChild(document.createTextNode(css));
  }

  return element;
}

function createInjector(style: HTMLStyleElement) {
  return new InjectorClient(style, createHashCounter(), createHashCounter(), createHashCounter());
}
