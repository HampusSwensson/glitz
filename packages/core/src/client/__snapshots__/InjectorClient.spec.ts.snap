// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client increments plain hash 1`] = `".c {color: blue;}"`;

exports[`client injects keyframes rule 1`] = `
"@keyframes a { 
  from {color: red;} 
  to {color: green;} 
}"
`;

exports[`client injects plain rule 1`] = `".a {color: red;}"`;

exports[`client injects plain rule 2`] = `".b {color: green; background-color: black;}"`;

exports[`client injects pseudo rule 1`] = `".a:hover {color: red;}"`;

exports[`client injects pseudo rule 2`] = `".b:hover {color: green; background-color: black;}"`;
